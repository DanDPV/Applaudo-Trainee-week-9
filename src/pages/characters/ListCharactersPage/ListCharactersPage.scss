@import '../../../styles/components/card';

$bg-title-image: 'https://wallpaperaccess.com/full/229965.jpg';

:root {
  --font-sans: 'Rubik', sans-serif;
}

.page-title-div {
  color: white;
  height: 250px;
  display: flex;
  justify-content: center;
  align-items: flex-end;
  padding-bottom: 20px;

  background: -moz-linear-gradient(top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0) 59%, rgba(0, 0, 0, 0.65) 100%), url($bg-title-image) no-repeat;
  background: -webkit-gradient(linear, left top, left bottom, color-stop(0%, rgba(0, 0, 0, 0)), color-stop(59%, rgba(0, 0, 0, 0)), color-stop(100%, rgba(0, 0, 0, 0.65))), url($bg-title-image) no-repeat;
  background: -webkit-linear-gradient(top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0) 59%, rgba(0, 0, 0, 0.65) 100%), url($bg-title-image) no-repeat;
  background: -o-linear-gradient(top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0) 59%, rgba(0, 0, 0, 0.65) 100%), url($bg-title-image) no-repeat;
  background: -ms-linear-gradient(top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0) 59%, rgba(0, 0, 0, 0.65) 100%), url($bg-title-image) no-repeat;
  background: linear-gradient(to bottom, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0) 59%, rgba(0, 0, 0, 0.65) 100%), url($bg-title-image) no-repeat;

  background-position: top;
}

.error-message {
  text-align: center;
  margin: 50px 15%;
}

.cards {
  display: grid;
  place-items: center;
}

.cards-content {
  display: grid;
  grid-gap: 1rem;
  padding: 1rem;
  max-width: 1024px;
  margin: 0 auto;
  font-family: var(--font-sans);

  @media (min-width: $bp-md) {
    grid-template-columns: repeat(2, 1fr);
  }

  @media (min-width: $bp-lg) {
    grid-template-columns: repeat(4, 1fr);
  }
}
