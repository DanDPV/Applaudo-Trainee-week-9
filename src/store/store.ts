import searchReducer from 'reducers/search/searchReducer';
import { createStore, combineReducers } from 'redux';

const rootReducer = combineReducers({
  search: searchReducer,
});

const store = createStore(rootReducer);

export default store;
