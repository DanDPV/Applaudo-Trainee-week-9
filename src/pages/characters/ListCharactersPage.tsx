import React, { useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IRootState } from 'store/store';
import IGenericApiResponse from 'interfaces/IGenericApiResponse';
import ICharacter from 'interfaces/ICharacter';
import useFetch from 'hooks/useFetch';

const ListCharactersPage = () => {
  const dispatch = useDispatch();
  const { url } = useSelector((state: IRootState) => state.search);
  const { loading } = useFetch<IGenericApiResponse<ICharacter>>(url);

  useEffect(() => {
    dispatch({
      type: 'SET_BASE_URL',
      payload: {
        baseUrl: `${process.env.REACT_APP_API_URL}v1/public/characters`,
      },
    });

    return () => {
      dispatch({
        type: 'RESET',
      });
    };
  }, []);

  return (
    <div className="main-content mb-5">
      <h1>List Characters Page</h1>
      <h2>{JSON.stringify(loading)}</h2>
    </div>
  );
};

export default ListCharactersPage;
