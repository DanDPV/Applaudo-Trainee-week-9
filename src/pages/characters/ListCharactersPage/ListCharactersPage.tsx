/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint no-shadow: 0 */
/* eslint no-unused-vars: 0 */
import React, { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import queryString from 'query-string';
import { useHistory } from 'react-router-dom';
import { debounce } from 'lodash';
import { IRootState } from 'store/store';
import IGenericApiResponse from 'interfaces/IGenericApiResponse';
import ICharacter from 'interfaces/ICharacter';
import useFetch from 'hooks/useFetch';
import Card from 'components/Card/Card';
import Loading from 'components/Loading/Loading';
import Pagination from 'components/Pagination/Pagination';
import { getQueryVariable } from 'utils/utils';
import {
  reset,
  setAllParams,
  setBaseUrl,
  setName,
} from 'actions/search';
import 'pages/characters/ListCharactersPage/ListCharactersPage.scss';

const ListCharactersPage = () => {
  enum QuerysParams {
    Page = 'page',
    Name = 'name',
  }

  const history = useHistory();
  const dispatch = useDispatch();
  const {
    url,
    limit,
    offset,
    name,
  } = useSelector((state: IRootState) => state.search);
  const { loading, data: genericResponse, error } = useFetch<IGenericApiResponse<ICharacter>>(url);

  const { data } = genericResponse ?? {};
  const { results, total } = data ?? {};

  const handleChangePage = (newPage: number) => history.push(
    `?${queryString.stringify({
      page: newPage,
      name,
    })}`,
  );

  const debouncedNameChange = useCallback(
    debounce((name: string) => history.push(
      `?${queryString.stringify({
        name,
        page: 1,
      })}`,
    ), 1000),
    [],
  );

  const handleNameChange = ({
    target,
  }: React.ChangeEvent<HTMLInputElement>) => {
    dispatch(setName(target.value));

    debouncedNameChange(target.value);
  };

  useEffect(() => {
    dispatch(setBaseUrl(`${process.env.REACT_APP_API_URL}v1/public/characters`));

    return () => {
      dispatch(reset());
    };
  }, []);

  useEffect(() => {
    const newPage = getQueryVariable(QuerysParams.Page)
      ? +getQueryVariable(QuerysParams.Page)
      : 1;

    const newName = getQueryVariable(QuerysParams.Name)
      ? decodeURI(getQueryVariable(QuerysParams.Name))
      : '';

    const newOffset = Math.ceil((newPage - 1) * limit);

    dispatch(setAllParams(newOffset, newName));
  }, [history.location, limit]);

  return (
    <div className="main-content mb-5">
      <div className="page-title-div">
        <h1>Characters</h1>
      </div>
      <div className="search-filters-form">
        <form>
          <label htmlFor="name">Name</label>
          <input
            type="text"
            name="name"
            autoComplete="off"
            value={name}
            onChange={handleNameChange}
          />
        </form>
      </div>
      {loading && <Loading />}
      {error && <h2 className="error-message">Could not load characters ðŸ˜“</h2>}
      <div className="cards">
        <div className="cards-content">
          {results
            && results.map(char => (
              <Card
                key={char.id}
                name={char.name}
                description={char.description}
                imageUrl={`${char.thumbnail.path}/portrait_uncanny.${char.thumbnail.extension}`}
              />
            ))}
        </div>
      </div>
      {results && results.length <= 0 && (
        <h2 className="error-message">Characters not found ðŸ˜®</h2>
      )}
      {!loading && !error && results && results.length > 0 && (
        <Pagination
          offset={offset}
          limit={limit}
          total={total ?? 0}
          onChange={handleChangePage}
        />
      )}
    </div>
  );
};

export default ListCharactersPage;
