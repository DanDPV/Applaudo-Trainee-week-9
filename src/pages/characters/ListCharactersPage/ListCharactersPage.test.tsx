/* eslint-disable no-unused-vars */
/* eslint-disable import/extensions */
/* eslint-disable no-undef */
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import store from 'store/store';
import ListCharactersPage from './ListCharactersPage';

describe('Test on ListCharactersPage component', () => {
  test('Should render component correctly', async () => {
    const { container } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ListCharactersPage />
        </BrowserRouter>
      </Provider>,
    );

    await waitFor(() => {
      expect(container.querySelector('.card')).not.toBeNull();
      expect(container.querySelector('.loading-container')).toBeNull();
      expect(container).toMatchSnapshot();
    });
  });
});
