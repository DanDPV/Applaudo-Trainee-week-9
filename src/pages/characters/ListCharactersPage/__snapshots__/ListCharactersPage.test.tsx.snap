// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on ListCharactersPage component Should render component correctly 1`] = `
<div>
  <div
    class="main-content mb-5"
  >
    <div
      class="page-title-div"
    >
      <h1>
        Characters
      </h1>
    </div>
    <div
      class="search-filters-form mb-5"
    >
      <form>
        <div
          class="search-title"
        >
          Search your hero
        </div>
        <div
          class="search-header"
        >
          Name
        </div>
        <div
          class="search-header"
        >
          Comic
        </div>
        <div
          class="search-header"
        >
          Story
        </div>
        <div
          class="search-value"
        >
          <input
            autocomplete="off"
            class="search-input"
            name="name"
            placeholder="HeroÂ´s name"
            type="text"
            value=""
          />
        </div>
        <div
          class="search-value"
        >
          <select
            class="search-select"
          >
            <option
              value=""
            >
              Select a comic
            </option>
            <option
              value="1"
            >
              comic spiderman
            </option>
          </select>
        </div>
        <div
          class="search-value"
        >
          <select
            class="search-select"
          >
            <option
              value=""
            >
              Select a story
            </option>
            <option
              value="1"
            >
              comic spiderman
            </option>
          </select>
        </div>
      </form>
    </div>
    
    <div
      class="cards"
    >
      <div
        class="cards-content"
      >
        <div
          class="card"
          style="background-image: url(path/portrait_uncanny.jpg);"
        >
          <div
            class="content"
          >
            <h2
              class="title"
            >
              spiderman
            </h2>
            <p
              class="copy"
            >
              description
            </p>
            <div
              class="card-actions"
            >
              <button
                class="btn btn-more"
                type="button"
              >
                View More
              </button>
              <button
                class="btn btn-action btn-hide"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-eye-slash fa-w-20 "
                  data-icon="eye-slash"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 640 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M320 400c-75.85 0-137.25-58.71-142.9-133.11L72.2 185.82c-13.79 17.3-26.48 35.59-36.72 55.59a32.35 32.35 0 0 0 0 29.19C89.71 376.41 197.07 448 320 448c26.91 0 52.87-4 77.89-10.46L346 397.39a144.13 144.13 0 0 1-26 2.61zm313.82 58.1l-110.55-85.44a331.25 331.25 0 0 0 81.25-102.07 32.35 32.35 0 0 0 0-29.19C550.29 135.59 442.93 64 320 64a308.15 308.15 0 0 0-147.32 37.7L45.46 3.37A16 16 0 0 0 23 6.18L3.37 31.45A16 16 0 0 0 6.18 53.9l588.36 454.73a16 16 0 0 0 22.46-2.81l19.64-25.27a16 16 0 0 0-2.82-22.45zm-183.72-142l-39.3-30.38A94.75 94.75 0 0 0 416 256a94.76 94.76 0 0 0-121.31-92.21A47.65 47.65 0 0 1 304 192a46.64 46.64 0 0 1-1.54 10l-73.61-56.89A142.31 142.31 0 0 1 320 112a143.92 143.92 0 0 1 144 144c0 21.63-5.29 41.79-13.9 60.11z"
                    fill="currentColor"
                  />
                </svg>
              </button>
              <button
                class="btn btn-action btn-bookmark"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-bookmark fa-w-12 "
                  data-icon="bookmark"
                  data-prefix="far"
                  focusable="false"
                  role="img"
                  viewBox="0 0 384 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M336 0H48C21.49 0 0 21.49 0 48v464l192-112 192 112V48c0-26.51-21.49-48-48-48zm0 428.43l-144-84-144 84V54a6 6 0 0 1 6-6h276c3.314 0 6 2.683 6 5.996V428.43z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="pagination-options"
    >
      <div
        class="multi-button"
      />
    </div>
    <div
      class="current-pagination-page"
    >
      <p>
        <b>
          Page: 
        </b>
        1 / 1
      </p>
    </div>
  </div>
</div>
`;
